Remember the status of our translations.

jfly opened the pull request:
This is safe to do because they'll only change when we restart the server.
This fixes #1270.

jfly commented on WcaOnRails/app/controllers/server_status_controller.rb:
```diff
@@ -2,6 +2,22 @@
 
+  def self.bad_keys_by_type_by_locale
+    @bad_keys_by_type_by_locale ||= begin
+      bad_keys_by_type_by_locale = {}
```
Still a Ruby noob here. We've got 3 different things called `bad_keys_by_type_by_locale`: the function we're in, the class instance variable `@bad_keys_by_type_by_locale`, and the local variable `bad_keys_by_type_by_locale`. Kind of crazy. @larspetrus, @jonatanklosko, I'm curious if you think this is okay, or if you'd approach this differently.

larspetrus replied:
I would boil it down to this:

    def self.bad_i18n_keys(locale)
      @bad_keys ||= {}.tap do |hash|
        ref_english = Locale.new('en')
        (I18n.available_locales - [:en]).each do |locale|
          missing, unused, outdated = Locale.new(locale, true).compare_to(ref_english)
          hash[locale] = { missing: missing, unused: unused, outdated: outdated }
        end
      end
      @bad_keys[locale]
    end

jfly replied:
I like that, although I'm not I'm not in love with the variable named `hash`. I also don't love that `bad_keys` sounds like an array, but I think that's my own problem I need to get over (sometimes I fear that my variable naming preferences have me on a path that ends in Hungarian notation).

Still, this solves a real problem, and I think it's better than my implementation, so I'm going to merge it.

jonatanklosko replied:
We could even get rid of the method, not sure if it's better, but I seeing your code I thought of `Hash.new` with a block.
```ruby
BAD_KEYS = Hash.new do |hash, locale|
  ref_english = Locale.new('en')
  (I18n.available_locales - [:en]).each do |locale|
    missing, unused, outdated = Locale.new(locale, true).compare_to(ref_english)
    hash[locale] = { missing: missing, unused: unused, outdated: outdated }
  end
end
```

larspetrus replied:
Not in love with `hash` either. I usually call these `result`, but it doesn't quite fit here. And I wanted to avoid yet another thing with the same name.

Yeah, describing the data structure returned in detail is rarely the best naming scheme. What the function is intended for is a better aim.

larspetrus replied:
Oh, Jonatan remembered that this is static data, so we can make it a constant!!

Yes, of course!

It should be `BAD_I18N_KEYS` in that case.

jonatanklosko replied:
Ah, actually there's no reason to iterate over them. I meant this of course:
```ruby
BAD_KEYS = Hash.new do |hash, locale|
  ref_english = Locale.new('en')
  missing, unused, outdated = Locale.new(locale, true).compare_to(ref_english)
  hash[locale] = { missing: missing, unused: unused, outdated: outdated }
end
```

jfly replied:
Good idea!

The problem with this is that `BAD_KEYS.keys` starts out empty, which would break our [`bad_keys` computation further down in the file](https://github.com/jfly/worldcubeassociation.org/blob/ca011bcdcbe29690ab218bdc4b12b20312ef52e7/WcaOnRails/app/controllers/server_status_controller.rb#L26).

It turns out that Rails has a nifty `each_with_object` that is nice here:

```ruby
BAD_KEYS_BY_TYPE_BY_LOCALE = (I18n.available_locales - [:en]).each_with_object({}) do |locale, hash|
  ref_english = Locale.new('en')
  missing, unused, outdated = Locale.new(locale, true).compare_to(ref_english)
  hash[locale] = { missing: missing, unused: unused, outdated: outdated }
end
```

Or if we don't want the dependency on RoR, Ruby [has a few native solutions](http://stackoverflow.com/questions/2943422/in-ruby-how-do-i-make-a-hash-from-an-array) for building up hashes:

```ruby
BAD_KEYS_BY_TYPE_BY_LOCALE = Hash[(I18n.available_locales - [:en]).map do |locale|
  ref_english = Locale.new('en')
  missing, unused, outdated = Locale.new(locale, true).compare_to(ref_english)
  bad_keys_by_type = { missing: missing, unused: unused, outdated: outdated }
  [ locale, bad_keys_by_type ]
end]
```

jonatanklosko commented:
LGTM! @larspetrus What do you think?

larspetrus commented:
1. I think this runs the first time the page is rendered, not at app startup. I don't want to slow down startup. But it doesn't, right?

2. Still dislike the naming. `BAD_KEYS_BY_TYPE_BY_LOCALE` contains minor details, but not the major fact that these are I18N keys. `TYPE` is meaningless to a reader not already knowing the structure, and he doesn't need to be told. The end use of `bad_keys_by_type = @bad_keys_by_type_by_locale[locale]` is mostly a bowl of word salad to me. Note how the `...by_locale[locale]` part is redundant and duplicative and also redundant.

So I'd go for `BAD_I18N_KEYS` and `@bad_i18n_keys`. Inline `@bad_i18n_keys[locale]` instead of creating `bad_keys_by_type`.

Then again, I wouldn't have mentioned that unless Jonatan asked, and if Jeremy wants to commit it this way, that's perfectly fine. LGTM.

jfly commented:
> I think this runs the first time the page is rendered, not at app startup. I don't want to slow down startup. But it doesn't, right?

My understanding is that it runs whenever the class loads. We do have `config.eager_load` turned on in [production.rb](https://github.com/jfly/worldcubeassociation.org/blob/bb5058665e5d9e7414a670083a4e86ffdceaecd6/WcaOnRails/config/environments/production.rb#L8-L12), so it's very possible this does happen on restart. I'll investigate this and report back. (If the answer is "yes, this does happen on startup", switching away from @jonatanklosko's constant and back to a memoized method would solve this problem.)

> BAD_KEYS_BY_TYPE_BY_LOCALE contains minor details, but not the major fact that these are I18N keys.

Good point, I totally missed that in my myopia. I am sold on renaming it to `BAD_I18N_KEYS`.

Will do one more tweak with some minor variable names and merge this up immediately.

Thanks for the feedback, all!

jfly commented:
> Will do one more tweak with some minor variable names and merge this up immediately.

Oh whoops, nevermind. I will not merge this up immediately because we want to figure out if this is happening on app startup or not.

jfly commented:
I added a big print statement to the code that generates `BAD_I18N_KEYS`. Here's what I saw:

Currently on server startup I see this:

```
~/gitting/worldcubeassociation.org/WcaOnRails @kaladin> bin/rails s
=> Booting WEBrick
=> Rails 4.2.7.1 application starting in development on http://localhost:3000
=> Run `rails server -h` for more startup options
=> Ctrl-C to shutdown server
[2017-02-02 11:41:02] INFO  WEBrick 1.3.1
[2017-02-02 11:41:02] INFO  ruby 2.3.3 (2016-11-21) [x86_64-linux]
[2017-02-02 11:41:02] INFO  WEBrick::HTTPServer#start: pid=28985 port=3000
```

Nothing happens until I do a query for `/server-status` in my browser, then you can see the BAD_I18N_KEYS get generated:

```
Started GET "/server-status" for ::1 at 2017-02-02 11:43:29 -0800
  ActiveRecord::SchemaMigration Load (0.4ms)  SELECT `schema_migrations`.* FROM `schema_migrations`
HIUAAAAAAA HERE I AM GENERATING BAD_I18N_KEYS
Processing by ServerStatusController#index as HTML
  Delayed::Backend::ActiveRecord::Job Load (0.4ms)  SELECT  `delayed_jobs`.* FROM `delayed_jobs` WHERE `delayed_jobs`.`attempts` = 0 AND (created_at < '2017-02-02 19:38:35.159471')  ORDER BY `delayed_jobs`.`created_at` ASC LIMIT 1
  Rendered server_status/index.html.erb within layouts/application (3.2ms)
  Rendered layouts/_navigation.html.erb (1061.4ms)
  Rendered layouts/_flash.html.erb (0.4ms)
  Rendered layouts/_footer.html.erb (0.3ms)
Completed 200 OK in 1438ms (Views: 1420.2ms | ActiveRecord: 0.8ms)
[User Id] Request was made by user id: <not logged in>
...
```

*Then* I changed `config.eager_load` to true in development.rb, and look what happened!

```
~/gitting/worldcubeassociation.org/WcaOnRails @kaladin> bin/rails s
=> Booting WEBrick
=> Rails 4.2.7.1 application starting in development on http://localhost:3000
=> Run `rails server -h` for more startup options
=> Ctrl-C to shutdown server
HIUAAAAAAA HERE I AM GENERATING BAD_I18N_KEYS
[2017-02-02 11:45:07] INFO  WEBrick 1.3.1
[2017-02-02 11:45:07] INFO  ruby 2.3.3 (2016-11-21) [x86_64-linux]
[2017-02-02 11:45:07] INFO  WEBrick::HTTPServer#start: pid=29832 port=3000
```

So, this will affect server startup. I don't actually have an opinion on how bad that is. @larspetrus, what do you think?

larspetrus commented:
I didn't know this is controlled by `config.eager_load`. This might explain why I'm confused about how this works - it depends!

If it only uses 0.1 sec in prod, I'm not really concerned. For development or test would have been worse. If it's 1.0 sec, I might start having doubts. Anyway, feel free to do what you want here to get this out and working!

PS. Possibly disruptive question: Why are we caching the data when we could just cache the HTML fragment?

jfly commented:
It looks like it's around ~7 seconds right now, actually. Here's how I timed it:

```ruby
  BAD_I18N_KEYS = begin
    start = Time.now#<<<
    (I18n.available_locales - [:en]).each_with_object({}) do |locale, hash| 
      ref_english = Locale.new('en')
      missing, unused, outdated = Locale.new(locale, true).compare_to(ref_english)
      hash[locale] = { missing: missing, unused: unused, outdated: outdated }
    end
    stop = Time.now#<<<
    puts "That took #{stop-start} seconds" #<<<
  end
```

```
~/gitting/worldcubeassociation.org/WcaOnRails @kaladin> bin/rails s
=> Booting WEBrick
=> Rails 4.2.7.1 application starting in development on http://localhost:3000
=> Run `rails server -h` for more startup options
=> Ctrl-C to shutdown server
That took 6.678672439 seconds
[2017-02-02 17:06:44] INFO  WEBrick 1.3.1
[2017-02-02 17:06:44] INFO  ruby 2.3.3 (2016-11-21) [x86_64-linux]
[2017-02-02 17:06:44] INFO  WEBrick::HTTPServer#start: pid=1150 port=3000
```

jfly commented:
> PS. Possibly disruptive question: Why are we caching the data when we could just cache the HTML fragment?

As I understand it, caching the HTML fragment wouldn't do much here. Generating the HTML doesn't take much time, it's that we're spending a bunch of time in the controller. I've never used it, but my understanding is that using [Rails's low-level caching](http://edgeguides.rubyonrails.org/caching_with_rails.html#low-level-caching) in the controller would help here, but then we'd have to worry about expiring/clearing that cache, as it would persist across server deploys.

larspetrus commented:
OK. At 7 secs I think it should be moved out of the way. That number will only increase.

Good point about how the HTML cache doesn't help. Two points really.

One downside of caching in the app is that each worker makes its own copy, instead of sharing one in Redis.

jfly commented:
I've changed it back to a memoized function. You can clearly see that the server starts up *without* doing this expensive computation.

```
~/gitting/worldcubeassociation.org/WcaOnRails @kaladin> bin/rails s
=> Booting WEBrick
=> Rails 4.2.7.1 application starting in development on http://localhost:3000
=> Run `rails server -h` for more startup options
=> Ctrl-C to shutdown server
[2017-02-02 22:07:05] INFO  WEBrick 1.3.1
[2017-02-02 22:07:05] INFO  ruby 2.3.3 (2016-11-21) [x86_64-linux]
[2017-02-02 22:07:05] INFO  WEBrick::HTTPServer#start: pid=13916 port=3000


Started GET "/" for ::1 at 2017-02-02 22:07:20 -0800
...
Started GET "/server-status" for ::1 at 2017-02-02 22:07:28 -0800
Processing by ServerStatusController#index as HTML
  Delayed::Backend::ActiveRecord::Job Load (0.4ms)  SELECT  `delayed_jobs`.* FROM `delayed_jobs` WHERE `delayed_jobs`.`attempts` = 0 AND (created_at < '2017-02-03 06:02:28.831875')  ORDER BY `delayed_jobs`.`created_at` ASC LIMIT 1
That took 5.562133196 seconds
  Rendered server_status/index.html.erb within layouts/application (2.0ms)
  Rendered layouts/_navigation.html.erb (6.2ms)
  Rendered layouts/_flash.html.erb (0.1ms)
  Rendered layouts/_footer.html.erb (0.1ms)
Completed 200 OK in 5676ms (Views: 110.1ms | ActiveRecord: 1.1ms)
[User Id] Request was made by user id: <not logged in>


Started GET "/server-status" for ::1 at 2017-02-02 22:07:34 -0800
Processing by ServerStatusController#index as HTML
  Delayed::Backend::ActiveRecord::Job Load (0.3ms)  SELECT  `delayed_jobs`.* FROM `delayed_jobs` WHERE `delayed_jobs`.`attempts` = 0 AND (created_at < '2017-02-03 06:02:34.521711')  ORDER BY `delayed_jobs`.`created_at` ASC LIMIT 1
That took 1.662e-06 seconds
  Rendered server_status/index.html.erb within layouts/application (1.4ms)
  Rendered layouts/_navigation.html.erb (2.7ms)
  Rendered layouts/_flash.html.erb (0.0ms)
  Rendered layouts/_footer.html.erb (0.1ms)
Completed 200 OK in 197ms (Views: 195.4ms | ActiveRecord: 0.3ms)
[User Id] Request was made by user id: <not logged in>








Started GET "/server-status" for ::1 at 2017-02-02 22:07:41 -0800
Processing by ServerStatusController#index as HTML
  Delayed::Backend::ActiveRecord::Job Load (0.3ms)  SELECT  `delayed_jobs`.* FROM `delayed_jobs` WHERE `delayed_jobs`.`attempts` = 0 AND (created_at < '2017-02-03 06:02:41.730625')  ORDER BY `delayed_jobs`.`created_at` ASC LIMIT 1
That took 9.92e-07 seconds
  Rendered server_status/index.html.erb within layouts/application (0.9ms)
  Rendered layouts/_navigation.html.erb (2.5ms)
  Rendered layouts/_flash.html.erb (0.1ms)
  Rendered layouts/_footer.html.erb (0.1ms)
Completed 200 OK in 90ms (Views: 88.6ms | ActiveRecord: 0.3ms)
[User Id] Request was made by user id: <not logged in>
```
